### ISSUE TITLE
HIR Visitor Refactoring

### ISSUE THREAD

**Issue by philberty** [Rust-GCC/gccrs#825]
# Background

Our HIR implementation is based off our AST slowly but surely we are cleaning this up to make the HIR much slimer such as no macros and methods are functions, all loops are loop expressions etc.

As part of this bootstrapping from our AST we only have a single Visitor interface that contains abstract methods for every type see: https://github.com/Rust-GCC/gccrs/blob/0024bc2f028369b871a65ceb11b2fddfb0f9c3aa/gcc/rust/hir/tree/rust-hir-visitor.h

The problem here is that when we are working with the full tree of the HIR we are forced to create visitor classes which contain all of the potential types even though the flow of the tree is always:

  Crate -> Items / types -> Stmts/Items/types -> Expressions / types

So even though we have a fully and strongly typed Intermediate representation to work with, we are not getting the full advantages of this yet. The major issue we have is that to avoid having to create overblown classes implementing every node in the graph we have these base classes: https://github.com/Rust-GCC/gccrs/blob/0024bc2f028369b871a65ceb11b2fddfb0f9c3aa/gcc/rust/backend/rust-compile-base.h

Which are starting to make the code quite messy to work with, the other problem is that every method has an empty block default we have cases in the code that do not have implementation such as the recent issue with ReferenceTypes on ImplBlocks such as:

```rust
impl<T> Foo for &T {}
```

This was due to the fact that impl blocks act on any HIR::TypeNoBounds and this visitor was missing an implementation for reference types:

https://github.com/Rust-GCC/gccrs/blob/0024bc2f028369b871a65ceb11b2fddfb0f9c3aa/gcc/rust/resolve/rust-ast-resolve-type.h#L125

Although that example is in the AST it shows how this visitor was missing making it very easy to miss cases in the full IR graph. If we have many smaller visitors we can do away with these base classes and use the C++ compiler to enforce the missing methods through using visitor abstract base classes. This will make it less confusing what methods we need to implement.

Then to take advantage of these new visitors let's focus this on the backend module (this should also be renamed to generic or something but that's a separate issue, backend is a bit of a weird name). If we focus usage of these new HIR visitors on the backend we limit the impact in finding regressions in the type-checking and lint passes until we are happy with this and the backend pass is a simpler case to start with.

https://github.com/Rust-GCC/gccrs/blob/0024bc2f028369b871a65ceb11b2fddfb0f9c3aa/gcc/rust/hir/tree/rust-hir-item.h

This single issue is not the finished goal but the starting point in a journey to make gccrs easier to work with and debug.

## Tasks

- [x] Refactor the existing HIR Visitor to be called HIRFullVisitor as shown here https://github.com/Rust-GCC/gccrs/commit/eb6963324cd48fb4774ad38723562b98e2d9dd28

This visitor will be different to the following visitors in that it will contain a non-abstract base class so that you can choose which methods we wish to override.

### HIRVisItemVisitor

rust-hir-item.h
- [ ] 697:class Module : public VisItem
- [ ] 776:class ExternCrate : public VisItem
- [ ] 1038:class UseDeclaration : public VisItem
- [ ] 1098:class Function : public VisItem, public ImplItem
- [ ] 1257:class TypeAlias : public VisItem, public ImplItem
- [ ] 1371:class Struct : public VisItem
- [ ] 1833:class Enum : public VisItem
- [ ] 1941:class Union : public VisItem
- [ ] 2028:class ConstantItem : public VisItem, public ImplItem
- [ ] 2111:class StaticItem : public VisItem
- [ ] 2503:class Trait : public VisItem
- [ ] 2627:class ImplBlock : public VisItem
- [ ] 3017:class ExternBlock : public VisItem

Every pass in the compiler starts with a top-level Item pass as these are the top-level constructs in the language

###  HIRImplItemisitor

The next step is ImplItems since when walking the graph we end up hitting ImplBlocks early which contains a list of ImplItems so we need to be able to access those:

rust-hir-item.h

- [ ] 1098:class Function : public VisItem, public ImplItem
- [ ] 1257:class TypeAlias : public VisItem, public ImplItem
- [ ] 2028:class ConstantItem : public VisItem, public ImplItem

### HIRTypeVisitor

Next we need to be able to access types in Rust we really have two different kind of type HIR::Type the overall type base class or we have TypeNoBounds. For now as part of this cleanup lets just start with a simple class for all types

rust-hir.h

- [ ] 377:class TypeNoBounds : public Type          // I think these might just be base classes
- [ ] 436:class Lifetime : public TypeParamBound // this is a types but we don't do anything with it

rust-hir-path.h

- [ ] 431:class TypePathSegmentGeneric : public TypePathSegment
- [ ] 566:class TypePathSegmentFunction : public TypePathSegment
- [ ] 611:class TypePath : public TypeNoBounds
- [ ] 845:class QualifiedPathInType : public TypeNoBounds

rust-hir-type.h

- [ ] 33:class TraitBound : public TypeParamBound
- [ ] 86:class ImplTraitType : public Type
- [ ] 145:class TraitObjectType : public Type
- [ ] 213:class ParenthesisedType : public TypeNoBounds
- [ ] 281:class ImplTraitTypeOneBound : public TypeNoBounds // this can probably go away in a future pr by desugaring into ImplTraitType but that's a seperate issue
- [ ] 320:class TupleType : public TypeNoBounds
- [ ] 388:class NeverType : public TypeNoBounds
- [ ] 417:class RawPointerType : public TypeNoBounds
- [ ] 484:class ReferenceType : public TypeNoBounds
- [ ] 559:class ArrayType : public TypeNoBounds
- [ ] 623:class SliceType : public TypeNoBounds
- [ ] 676:class InferredType : public TypeNoBounds
- [ ] 785:class BareFunctionType : public TypeNoBounds

### HIRStmtVisitor

This one is a bit more annoying since Stmts can also contain some Items but let's just handle it like the others

rust-hir-stmt.h

- [ ] 29:class EmptyStmt : public Stmt
- [ ] 54:class LetStmt : public Stmt
- [ ] 134:class ExprStmt : public Stmt
- [ ] 153:class ExprStmtWithoutBlock : public ExprStmt
- [ ] 197:class ExprStmtWithBlock : public ExprStmt

rust-hir.h 
119:class Item : public Stmt

// ^^ this hierarchy is a bit annoying at we may want to change it in the future because of enum-item making it confusing so let's just get it to contain the relevant VisItems for now

- [ ] 697:class Module : public VisItem
- [ ] 776:class ExternCrate : public VisItem
- [ ] 1038:class UseDeclaration : public VisItem
- [ ] 1098:class Function : public VisItem, public ImplItem
- [ ] 1257:class TypeAlias : public VisItem, public ImplItem
- [ ] 1371:class Struct : public VisItem
- [ ] 1833:class Enum : public VisItem
- [ ] 1941:class Union : public VisItem
- [ ] 2028:class ConstantItem : public VisItem, public ImplItem
- [ ] 2111:class StaticItem : public VisItem
- [ ] 2503:class Trait : public VisItem
- [ ] 2627:class ImplBlock : public VisItem
- [ ] 3017:class ExternBlock : public VisItem

This is all because Rust allows for Blocks to contain items functions etc see: https://doc.rust-lang.org/reference/statements.html I wonder are there any limits on statements within blocks the reference does not mention any but maybe we will need someone to double-check this at some point we might be able to remove some things from this such as the extern block's maybe I have never seen those within a rust block expression.

### HIRExpressionVisitor

Finally lets finish this initial refactor with an expression visitor:

- [ ] rust-hir.h
- [ ] 225:class ExprWithoutBlock : public Expr
- [ ] 266:class IdentifierExpr : public ExprWithoutBlock
- [ ] 734:class PathExpr : public ExprWithoutBlock
- [ ] rust-hir-expr.h
- [ ] 32:class ExprWithBlock : public Expr // this is a base class
- [ ] 61:class LiteralExpr : public ExprWithoutBlock
- [ ] 125:class OperatorExpr : public ExprWithoutBlock
- [ ] 744:class GroupedExpr : public ExprWithoutBlock
- [ ] 933:class ArrayExpr : public ExprWithoutBlock
- [ ] 1008:class ArrayIndexExpr : public ExprWithoutBlock
- [ ] 1073:class TupleExpr : public ExprWithoutBlock
- [ ] 1157:class TupleIndexExpr : public ExprWithoutBlock
- [ ] 1228:class StructExpr : public ExprWithoutBlock
- [ ] 1628:class CallExpr : public ExprWithoutBlock
- [ ] 1708:class MethodCallExpr : public ExprWithoutBlock
- [ ] 1796:class FieldAccessExpr : public ExprWithoutBlock
- [ ] 1920:class ClosureExpr : public ExprWithoutBlock
- [ ] 2002:class BlockExpr : public ExprWithBlock
- [ ] 2181:class ContinueExpr : public ExprWithoutBlock
- [ ] 2224:class BreakExpr : public ExprWithoutBlock
- [ ] 2301:class RangeExpr : public ExprWithoutBlock
- [ ] 2609:class ReturnExpr : public ExprWithoutBlock
- [ ] 2677:class UnsafeBlockExpr : public ExprWithBlock
- [ ] 2765:class BaseLoopExpr : public ExprWithBlock
- [ ] 3053:class IfExpr : public ExprWithBlock
- [ ] 3266:class IfLetExpr : public ExprWithBlock
- [ ] 3840:class MatchExpr : public ExprWithBlock
- [ ] 3919:class AwaitExpr : public ExprWithoutBlock
- [ ] 3969:class AsyncBlockExpr : public ExprWithBlock

## TLDR on Tasks:

Let's focus the initial refactor on these 5 new visitors and use them to update the code within the Backend module for now.

- HIRVisItemVisitor
- HIRImplItemisitor
- HIRTypeVisitor
- HIRStmtVisitor
- HIRExpressionVisitor

**Comment by bjorn3** [comment]
> Crate -> Items / types -> Stmts/Items/types -> Expressions / types

You can put items inside expressions. For example serde does something like

```rust
const _: () = {
    use serde::Serialize;
    impl Serialize for MyType {
        // ...
    }
};
```

to provide a new scope that doesn't leak `use`d items.

**Comment by philberty** [comment]
Ah yes i missed that in that comment but our pipeline currently allows for that right now because of the HIRStmtVisitor task I also referenced to https://doc.rust-lang.org/reference/statements.html

But I wonder are there any limits on using items within blocks? We currently support types and functions so far but need this refactor work to reuse more of our existing code for the others

**Comment by bjorn3** [comment]
I don't think there are limits. Even module includes work inside a function for some arcane reason.

**Comment by dkm** [comment]
Small update: I've finally got a working compiler (i.e. it builds) and surprisingly, the first testsuite run is fine : 
```
                === rust Summary ===

# of expected passes            720

                === rust Summary ===

# of expected passes            922
# of expected failures          2

                === rust Summary ===

# of expected passes            834
# of expected failures          7

                === rust Summary ===

# of expected passes            1175
# of expected failures          5

                === rust Summary ===

# of expected passes            972
# of expected failures          7

                === rust Summary ===

# of expected passes            1062
```
I'll still have a lot to clean, but I think the hardest part is done.

**Comment by dkm** [comment]
@philberty I have a first version. I have to verify the list of nodes supported by the different visitors, but it should already be mostly ok. Currently, it does the same thing as before:
- abstract visitor classes
- ... basic classes implem with empty behavior for all nodes
- specific classes inherit from basic classes and override what they need

I think that could be the next step for this refactor. Then, in a following step, I should remove the basic classes, pushing all the work in the specific classes, as that was the initial goal (be more explicit on what the specific class should handle).

What do you think ?

**Comment by dkm** [comment]
> I think that could be the next step for this refactor. Then, in a following step, I should remove the basic classes, pushing all the work in the specific classes, as that was the initial goal (be more explicit on what the specific class should handle).

I did the extra change to have only the abstract visitor and the impl class. Makes the visitor classes having only few `visit` to implement a bit more verbose because they are forced to provide empty behavior for the unused nodes...
